File: 0061-rotate-list/0061-rotate-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var rotateRight = function(head, k) {
    if (!head) return head;
    let [curr, len, i] = [head, 0, 0];
    while (curr) {
        len++;
        curr = curr.next;
    }
    const headIdx = (len - (k % len)) % len;
    if (headIdx === 0) return head;
    curr = head;
    while (i < headIdx - 1) {
        curr = curr.next;
        i++;
    }
    const last = curr;
    const newHead = curr.next;
    last.next = null;
    curr = newHead;
    while (curr.next) {
        curr = curr.next;
    }
    curr.next = head;
    return newHead;
};

File: 0061-rotate-list/0061-rotate-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *   Val int
 *   Next *ListNode
 * }
 */
func rotateRight(head *ListNode, k int) *ListNode {
  if head == nil {
    return head
  }
  curr, l, i := head, 0, 0
  for curr != nil {
    curr = curr.Next
    l++
  }
  headIdx := (l-(k%l))%l
  if headIdx == 0 {
    return head
  }
  curr = head
  for i < headIdx-1 {
    curr = curr.Next
    i++
  }
  last, newHead := curr, curr.Next
  last.Next, curr = nil, newHead
  for curr.Next != nil {
    curr = curr.Next
  }
  curr.Next = head
  return newHead
}